src/cast_core/src/CommitLatency.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright 2019-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/CommitLatency.hpp>
+
+#include <memory>
+
+#include <yaml-cpp/yaml.h>
+
+#include <bsoncxx/json.hpp>
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+#include <mongocxx/database.hpp>
+
+#include <boost/log/trivial.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/ExecutionStrategy.hpp>
+#include <gennylib/InvalidConfigurationException.hpp>
+#include <gennylib/context.hpp>
+#include <value_generators/value_generators.hpp>
+
+
+namespace genny::actor {
+
+struct CommitLatency::PhaseConfig {
+    mongocxx::collection collection;
+    mongocxx::write_concern wc;
+    mongocxx::read_concern rc;
+    mongocxx::read_preference rp;
+    std::string rp_string;
+    std::shared_ptr<mongocxx::client_session> session;
+    mongocxx::options::find optionsFind;
+    mongocxx::options::update optionsUpdate;
+    mongocxx::options::aggregate optionsAggregate;
+    bool useSession;
+    int64_t repeat;
+    int64_t threads;
+    std::uniform_int_distribution<int64_t> amountDistribution;
+    ExecutionStrategy::RunOptions options;
+
+    PhaseConfig(PhaseContext& phaseContext, const mongocxx::database& db)
+        : collection{db[phaseContext.get<std::string>("Collection")]},
+          rp_string{phaseContext.get<std::string>("ReadPreference")},
+          useSession{phaseContext.get<bool, false>("Session")},
```

Comment:
This syntax may not be doing what you think and I had to stare at it for a bit myself - the `get<T,false>` form returns a `std::optional`. This is effectively casting the `std::optional` to a bool by using an implicit-conversion. The `std::optional<boolean>` will itself evaluate to `true` _iff_ it's populated (if the user specified a value for the key). So if the user specifies `Session: false` in the YAML the `useSession` variable will be true. I think what you want here is

```c++
useSession{phaseContext.get<bool, false>("Session").value_or(false)}
```

Similar applies to elswhere here where this does `.get<bool,false>`

I see you did the `value_or` thing for a few others so this could be intentional (in which case may be worthy of a comment).

Reply:
I understood the functionality of std::optional such that if my default is false, then the `value_or(false)` is redundant. But I can add them for clarity if that's our style. Clarity is always nice.